Sming/SmingCore/Network/HttpServer.cpp

```diff
@@ -85,21 +93,43 @@ void HttpServer::setDefaultHandler(const HttpPathDelegate& callback)
 	addPath("*", callback);
 }
 
-void HttpServer::addPath(const String& path, const HttpResourceDelegate& onRequestComplete) {
+void HttpServer::addPath(const String& path, const HttpResourceDelegate& onRequestComplete)
+{
 	HttpResource* resource = new HttpResource;
 	resource->onRequestComplete = onRequestComplete;
 	resourceTree[path] = resource;
 }
 
-void HttpServer::addPath(const String& path, HttpResource* resource) {
+void HttpServer::addPath(const String& path, HttpResource* resource)
+{
 	resourceTree[path] = resource;
 }
 
-void HttpServer::setDefaultResource(HttpResource* resource) {
+void HttpServer::setDefaultResource(HttpResource* resource)
+{
 	addPath("*", resource);
 }
 
-void HttpServer::onConnectionClose(TcpClient& connection, bool success) {
-	totalConnections--;
+void HttpServer::shutdown()
+{
+	active = false;
+	for(int i=0; i < connections.count(); i++) {
+		HttpServerConnection* connection = connections[i];
+		if(connection == NULL) {
+			continue;
+		}
+
+		connection->close();
+	}
+}
+
+void HttpServer::onConnectionClose(TcpClient& connection, bool success)
+{
+	connections.removeElement((HttpServerConnection*)&connection);
+	totalConnections = connections.count();
+	if(totalConnections == 0 && !active){
+		debugf("Shutting down the Http Server");
+		delete this;
```

Comment:
Note to myself: That will cause a race condition - HttpServer will start freeing the resources and meanwhile the HttpServerConnections will try to use those resource in their destructors... A timer with 3 - 4 seconds delay should fix this...